rollout/trafficrouting/alb/alb.go

```diff
@@ -0,0 +1,144 @@
+package alb
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	extensionslisters "k8s.io/client-go/listers/extensions/v1beta1"
+	"k8s.io/client-go/tools/record"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/diff"
+	ingressutil "github.com/argoproj/argo-rollouts/utils/ingress"
+	jsonutil "github.com/argoproj/argo-rollouts/utils/json"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+const (
+	// Type holds this controller type
+	Type = "ALB"
+)
+
+// ReconcilerConfig describes static configuration data for the ALB Ingress reconciler
+type ReconcilerConfig struct {
+	Rollout        *v1alpha1.Rollout
+	Client         kubernetes.Interface
+	Recorder       record.EventRecorder
+	ControllerKind schema.GroupVersionKind
+	IngressLister  extensionslisters.IngressLister
+}
+
+// Reconciler holds required fields to reconcile ALB Ingress resources
+type Reconciler struct {
+	cfg ReconcilerConfig
+	log *logrus.Entry
+}
+
+// NewReconciler returns a reconciler struct that brings the ALB Ingress into the desired state
+func NewReconciler(cfg ReconcilerConfig) *Reconciler {
+	return &Reconciler{
+		cfg: cfg,
+		log: logutil.WithRollout(cfg.Rollout).WithField(logutil.IngressKey, cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.ALB.Ingress),
+	}
+}
+
+// Type indicates this reconciler is an ALB ingress reconciler
+func (r *Reconciler) Type() string {
+	return Type
+}
+
+// Reconcile modifies ALB Ingress resources to reach desired state
+func (r *Reconciler) Reconcile(desiredWeight int32) error {
+	rollout := r.cfg.Rollout
+	ingressName := rollout.Spec.Strategy.Canary.TrafficRouting.ALB.Ingress
+	ingress, err := r.cfg.IngressLister.Ingresses(rollout.Namespace).Get(ingressName)
+	if err != nil {
+		return err
+	}
+	stableService := r.cfg.Rollout.Spec.Strategy.Canary.StableService
+	port := r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.ALB.ServicePort
+	if !ingressutil.HasRuleWithService(ingress, stableService) {
```

Comment:
We error here if there is no rule mentioning the service. But how does this work initially, when the ingress doesn't have the annotation yet?